File: 00-introduction/presidents.py
# From Digital Dice, page 7
import random

m = 24
totalcorrect = 0
m_range = range(0, m)
iterations = 1000000 # Python (or my implementation) is pretty slow with a million iterations.

for k in range(1, iterations):
    correct = 0
    random.shuffle(m_range)

    for j in range(0, m):
        if m_range[j] == j:
            correct += 1

    totalcorrect += correct

avg = totalcorrect * 1.0 / iterations

print "avg: ", avg

File: 00-introduction/presidents.js
var shuffle = function (o) {
    for (var j, x, i = o.length; i; j = parseInt(Math.random() * i, 10), x = o[--i], o[i] = o[j], o[j] = x);
    return o;
};

var m = 24;
var totalcorrect = 0;
var m_range = [];
for (var i = 0; i < 24; i++) {
    m_range.push(i);
}

var iterations = 1000000;
var correct = 0;

for (var k = 0; k < iterations; k++) {
    correct = 0;
    var shuffled = shuffle(m_range);

    for (var j = 0; j < m_range.length; j++) {
        if (shuffled[j] === j) {
            correct = correct + 1;
        }
    }

    totalcorrect = totalcorrect + correct;
}

var avg = totalcorrect / iterations;

console.log("avg: " + avg);
